src/drawer.canvasentry.js

```diff
@@ -159,26 +163,40 @@ export default class CanvasEntry {
     }
 
     /**
-     * Utility function to handle wave color arguments
+     * Utility function to determine the type and handle the argument passed to
+     * a color setting method
      *
-     * When the color argument type is a string, it will be returned as is.
-     * Otherwise, it will be treated as an array, and a canvas gradient will
-     * be returned
+     * Return type is dependent on the color argument's type
      *
-     * @since 5.3.0
+     * @since 6.0.0
      * @param {CanvasRenderingContext2D} ctx Rendering context of target canvas
-     * @param {string|string[]} color Fill color for the wave canvas, or an array of colors to apply as a gradient
-     * @returns {string|CanvasGradient} Returns a string fillstyle value, or a canvas gradient
+     * @param {string|string[]|CanvasImageSource} color A CSS color value as a
+     * string, an array of CSS color value strings, or a Canvas Image Source
+     * (HTML Image, SVG Image, HTML Video, HTML Canvas, Image Bitmap or
+     * Offscreen Canvas element)
+     * @returns {string|CanvasGradient|CanvasPattern} Returns a CSS color
+     * value string, a canvas gradient or a canvas pattern
      */
     getFillStyle(ctx, color) {
-        if (typeof color == 'string') {
+        // if the color argument is a string, handle it as a CSS color value
+        if (typeof color === 'string') {
             return color;
         }
 
-        const waveGradient = ctx.createLinearGradient(0, 0, 0, ctx.canvas.height);
-        color.forEach((value, index) => waveGradient.addColorStop((index / color.length), value));
+        // if it is an array, handle it as a gradient
+        if (Array.isArray(color)) {
+            const waveGradient = ctx.createLinearGradient(0, 0, 0, ctx.canvas.height);
+            color.forEach((value, index) => waveGradient.addColorStop((index / color.length), value));
+            return waveGradient;
+        }
+
+        // if it is an object, handle it as a CanvasImageSource
+        if (typeof color === "object") {
+            return ctx.createPattern(color, "repeat");
```

Comment:
now that we have new methods, let's think about allowing user to change "repeat" value (and other settings in new method). Thoughts?

Reply:
Add an optional options parameter to setWaveColor, setWaveStyle, etc.. that could include options depending on the type of style being set.  If it's a gradient, options for controlling the angle of the gradient.  If it's a Canvas Image Source, the repeat value.  

`setWaveStyle(myCanvImgSrc, {repeat: 'repeat-x'});`

Are you looking all the way back up to the WaveSurfer parameters object?  Or just up to the new methods?

Reply:
> Are you looking all the way back up to the WaveSurfer parameters object? Or just up to the new methods?

Yep, both.

Reply:
My recommendation is a new Wavesurfer Params object property called waveFillStyle, which would be an object type that could also be plugged into progressFillStyle, and eventually the wave background, the playhead, etc..)

`
```
  generateWaveform() {
    ...,
    waveFillStyle: {
      
      // only one of these (color, gradient, image, canvas, video) would be required
      color: {'red'},

      // each type of fill style would have its options as siblings of its data
      gradient: {
          colorstops: [
              [0.0,'red'],
              [0.5,'blue'],
              [1.0,'purple']
          ],
          coordinates: [0,0,500,500]
      },
      image: {
          src: document.querySelector("#image-element"),
          repeat: 'repeat'
      },

      // not yet, but eventually could pass any canvas as the style
      canvas: {
          element: document.querySelector("#canvas-element")
      },

      // I was able to pass in a canvas video as the background, it is pretty great,
      // but was quite a bit of setup, as in having to run an interval to update the canvas
      // from the video element, so still sorting the last pieces of that out before I 
      // push anything
      video: {
          element: document.querySelector("#video-canvas")
      }

  }`

And keeping waveColor, etc.. as is for backwards compatibility.

What do you think?

Reply:
Let's start with `setWaveStyle(myCanvImgSrc, {repeat: 'repeat-x'});` like you suggested. The object passed to `setWaveStyle` can be called `options`.

Reply:
In setWaveStyle, the style is being set to the waveColor parameter, and then drawBuffer is called.  How do I get access to this options object over in drawer.canvasentry.js's getFillStyle method?  

From what I can tell, I will need to 

- set the options object as a params property
- add that property as an argument to drawer.multicanvas.js's call to setFillStyles line 556
- add the wave and progressStyleOptions parameter to drawer.canvasentry.js's setFillStyles method line 156
- pass the new arguments to the getFillStyle calls lines 158 and 161
- subsequently pass it getFillStyle's ctx.createPattern call, where the repeat option can be utilized.

Is this what you had in mind?  It is what I am committing right now, and it works.

If this is not on track with what you were looking for, I will need more details on how else to make this happen.  Let me know!  Thanks